Expand inputs in processors for VLMs

zucchini-nlp opened the pull request:
# What does this PR do?

Fixes #30809, This PR moves the `_merge_inputs_with_vision_embeds` to the processing logics, and thus making VLMs more versatile in terms of generation strategies. All models were tested locally with different batch sizes and img resolutions, the generation is same as it was before making changes.

The main idea is to get sequence length for image features inside the processing files, and expand input ids by repeating special image token. Same is already done for IDEFICS in `transformers`.

zucchini-nlp committed:
let it be

zucchini-nlp committed:
draft

zucchini-nlp committed:
should not have changed

zucchini-nlp commented on src/transformers/models/llava_next/processing_llava_next.py:
```diff
@@ -108,12 +114,74 @@ def __call__(
 
         return BatchFeature(data={**text_inputs, **image_inputs})
 
+    def _get_number_of_features(self, height: int, width: int) -> int:
```
Mostly copied from TGI with minor changes in calculations for unpadding, otherwise it won't work for low resolution images

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_30962). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

molbap commented:
Looking forward to see this expanded to other VLMs! Some might be trickier, PaliGemma incorporates causal mask computation in the merge method for instance (thought about that when reading) but it makes sense that most should belong in the processor, not the modeling

zucchini-nlp committed:
add warnings

zucchini-nlp committed:
Merge remote-tracking branch 'upstream/main' into vlm_processors

zucchini-nlp committed:
fix & add tests

zucchini-nlp committed:
fix tests

zucchini-nlp commented:
@amyeroberts I did some clean-up after Arthur's comments. Requesting review, should be ready. If this works I will expand the logic to BLIP and PaliGemma in the next weeks 

What changed:
- Model can generate from both: old inputs and new-expanded inputs. If it's old inputs, warning is raised, asking to upgrade the processor config.
- Processor also can return both types. If it has all the necessary parts to calculate image embedding length, the inputs will be expanded. Otherwise, warning is raised and old behavior retained.
- Old behavior is planned to be totally removed in v4.44 (or better v4.43?)
- Added tests to check that old vs new inputs generation is identical
- To actually have llava-based models work in new style, I'll later update all hf-llava configs in the hub. Other models in the hub will continue to work with old behavior

zucchini-nlp committed:
ipnuts embeds cannot be passed with pixels

amyeroberts commented on src/transformers/models/llava/processing_llava.py:
```diff
@@ -45,6 +48,10 @@ class LlavaProcessor(ProcessorMixin):
 
     def __init__(self, image_processor=None, tokenizer=None):
         super().__init__(image_processor, tokenizer)
+        self.image_size = self.image_processor.size["shortest_edge"]
```
As llava can take any image processor, we can't assume `size` will have `shortest_edge`, it should also be a dict of `{"height": h, "width": w}`

zucchini-nlp replied:
since we can't be sure which one is used by image processor and what's the output size, I decided to check `pixel_values` size and use that as `self.image_size`

Also, now we don't assume it's always a square size output and calculate patches for height and width separately, similar to what ViT does in its modeliing file.

amyeroberts commented on src/transformers/models/llava/processing_llava.py:
```diff
@@ -45,6 +48,10 @@ class LlavaProcessor(ProcessorMixin):
         super().__init__(image_processor, tokenizer)
+        self.image_size = self.image_processor.size["shortest_edge"]
+        self.patch_size = getattr(self.image_processor, "patch_size", None)
+        self.image_token = "<image>"
```
Is this consistent for all the tokenizers?

zucchini-nlp replied:
For tokenizers, yes! The only part from tokenizer is "image token" which I believe is established as `"<image>"` for LLaVas.

For all image processors/vision towers, then maybe no. Calculation of image token seq length is done assuming we get CLIP as image tower. I am not sure if other vision backbones are all similar or require different handling. 

In case there are llavas w/o CLIP and it has to be handled differently, I will see if we can unify it without extra code-hacks

zucchini-nlp replied:
For existing possible backbones (CLIP, SigLIP, Dino, ViT) the current code will work as expected. Same for tokenizers, and in case any model decided to change image token, the user can set it by `processr.image_token="{ANY_NEW_TOKEN}"`

amyeroberts commented on src/transformers/models/llava/processing_llava.py:
```diff
@@ -104,10 +111,38 @@ def __call__(
+                prompt_strings.append(sample)
+        else:
+            prompt_strings = text
+            logger.warning_once(
+                "Expanding inputs for image tokens in LLaVa should be done in processing. "
+                "Please add `patch_size` and `vision_feature_select_strategy` to the model's image processing config. "
+                "Using processors without these attributes in the config is deprecated and will throw an error in v4.44."
+            )
```
I think this is both going to be annoying to handle and creates leakiness in our abstractions: not everyone who uses the image processor will be using it for llava, and there's no reason for image processors for models which don't have `patch_size` to have this attribute. 

Instead, I'd suggest getting the defaults from the image processor if available, but also allowing them as processor arguments in the init and possibly the forward pass which the user can directly set and control. This way, when we save out the processor, these can be saved alongside, and the user doesn't need to change an image processor or override the processor each time they want to use it

zucchini-nlp replied:
Right, that will be easier for users. Will make changes

amyeroberts commented on tests/models/llava/test_modeling_llava.py:
```diff
@@ -186,6 +186,48 @@ def setUp(self):
+            inputs["inputs_embeds"] = wte(input_ids)
+
+            with torch.no_grad():
+                model(**inputs)[0]
```
Why index on the 0'th output here ?

zucchini-nlp replied:
no reason, it was simply copied as is from `test_modeling_common.py` with extra handling of `pixel_values` as input to forward. 

I'll remove it here, as it's not making any difference

amyeroberts reviewed (commented):
Thanks for working on this - will be great to have some of this logic unified! 

Main comment is about how we set the required arguments for processing in the processor

zucchini-nlp committed:
more updates

zucchini-nlp committed:
paligemma ready!

zucchini-nlp committed:
minor typos

zucchini-nlp committed:
update blip-2

zucchini-nlp committed:
fix tests & raise error

zucchini-nlp committed:
Merge branch 'main' into vlm_processors

zucchini-nlp committed:
docstring

zucchini-nlp committed:
add blip2 test

zucchini-nlp committed:
Merge branch 'main' into vlm_processors

zucchini-nlp commented:
@amyeroberts addressed the comments and added all VLMs to the PR (excluding Idefics, Fuyu and Kosmos as those already have expansion in processing).

- warning text is more clear and it's easy for users to add new attributes to `Processor` class (with `processor.patch_size = patch_size`)
- BLIP-2 needed more modifications as it didn't have special image token, lmk if the way I did works
- Paligemma worked out of the box but needed changes for causal mask. There's also smth weird with "position_ids" which will be fixed by @molbap 
-  All models have their equivalence tests and are passing locally. I don't know how to make happy the failing doctest, it's red even after I deprecated the unused attribute

zucchini-nlp committed:
tmp

amyeroberts commented on src/transformers/models/vipllava/modeling_vipllava.py:
```diff
@@ -418,26 +418,48 @@ def forward(
+            # if the number of image tokens is more than image embeddings seq length, then prob we expanded it in processing
+            # not very reliable, but we don't expect one to actually pass 500+ images for one prompt
+            # In case we're in decoding stage, legacy behavior is checked by presence of pixel values even if use_cache=True
+            legacy_processing = ((input_ids == self.config.image_token_index).sum(1).max() < 576) or (
```
Is the 576 the sequence length? We should take this from the config, rather than hard-code here

zucchini-nlp replied:
Yes, it's the length for one image, and can be added as image_seq_length to all configs. I will start updating llava (and maybe others by opening a PR) configs today so the default behavior becomes the "expand in processor"

amyeroberts reviewed (commented):
Wow - a big piece of work! 

Overall looks good to me, just a few comments here and there. I'd like to have a second review from @molbap and a run on the slow tests for all the models touched here

zucchini-nlp commented on src/transformers/models/vipllava/modeling_vipllava.py:
```diff
@@ -418,26 +418,48 @@ def forward(
+                "You cannot specify both input_ids and inputs_embeds at the same time, and must specify either one"
+            )
+
+        if pixel_values is not None and inputs_embeds is not None:
+            raise ValueError(
+                "You cannot specify both pixel_values and inputs_embeds at the same time, and must specify either one"
+            )
```
Both can be `None`, that happens when we are in decoding stage and get only the last token for generation. We only can't have both set to a tensor, because that means we don't know where are `image_token` so we can't merge images into embeddings

zucchini-nlp commented on src/transformers/models/video_llava/processing_video_llava.py:
```diff
@@ -37,13 +40,34 @@ class VideoLlavaProcessor(ProcessorMixin):
+        image_token="<image>",  # set the default and let users change if they have peculiar special tokens in rare cases
+        video_token="<video>",
+    ):
+        self._patch_size = patch_size
+        self._vision_feature_select_strategy = vision_feature_select_strategy
```
oh, forgot to get it back, was part of an older logic

zucchini-nlp committed:
add image seq length to config

zucchini-nlp committed:
update docstring

zucchini-nlp committed:
Merge branch 'main' into vlm_processors

zucchini-nlp committed:
delete

zucchini-nlp committed:
fix tests

zucchini-nlp committed:
fix blip

molbap commented on src/transformers/models/llava/modeling_llava.py:
Same comment as for `modeling_paligemma`, wondering if it's an issue on my end? after the first token generated, `modeling_llama` will throw because of 
```python
AttributeError: 'tuple' object has no attribute 'get_seq_length'
``` (this does not happen on `main`)

molbap commented on src/transformers/models/llava/modeling_llava.py:
```diff
@@ -415,63 +419,89 @@ def forward(
-                # this is not memory efficient at all (output_hidden_states=True) will save all the hidden stated.
-                selected_image_feature = image_outputs.hidden_states[vision_feature_layer]
+            # if the number of image tokens is more than image embeddings seq length, then prob we expanded it in processing
+            # not very reliable, but we don't expect one to actually pass 500+ images for one prompt
```
couldn't we also have a very tiny image embeddings sequence length?

zucchini-nlp replied:
For LLaVa models the usual backbone is CLIP (i don't remember which ckpt exactly), which always give 576 length. In case some LLaVas have another vision backbone, I guess user have to switch the `image_seq_length` parameter in model's config. 

I am thinking if we can catch that and give user an informative error msg? Will explore more this issue

molbap reviewed (commented):
Will continue reviewing (biig work) but sharing feedback on the 2 first things I tested, Llava and Paligemma

zucchini-nlp commented on src/transformers/models/paligemma/modeling_paligemma.py:
Hmm, weird since we should be doing Cache class inside generate before going to the modeling part. I will check out tomorrow on this

zucchini-nlp replied:
Found the bug, PaliGemma didn't have the cache flags on. Resolved!

zucchini-nlp committed:
fix paligemma

zucchini-nlp committed:
merge `main`

zucchini-nlp committed:
out-of-place scatter

zucchini-nlp committed:
add llava-next-video

molbap reviewed (commented):
It's a biiig piece of work, nicely done, tests and all! I left a few comments on some things I didn't understand well + paligemma masking in particular

zucchini-nlp commented on src/transformers/models/blip_2/modeling_blip_2.py:
```diff
@@ -1767,12 +1767,25 @@ def forward(
+            logger.warning_once(
+                "Expanding inputs for image tokens in BLIP-2 should be done in processing. "
+                "Please follow instruction here (https://gist.github.com/zucchini-nlp/e9f20b054fa322f84ac9311d9ab67042) to update your BLIP-2 model. "
+                "Using processors without these attributes in the config is deprecated and will throw an error in v4.44."
```
Yes, will update accordingly when we get one step away from merging. I think two-three major versions from current one will work :)

zucchini-nlp committed:
Update src/transformers/models/blip_2/modeling_blip_2.py

Co-authored-by: Pablo Montalvo <39954772+molbap@users.noreply.github.com>

zucchini-nlp commented on src/transformers/models/blip_2/processing_blip_2.py:
```diff
@@ -121,9 +143,27 @@ def __call__(
+                image_tokens = self.image_token.content * self.num_query_tokens
+                image_token_encoding = self.tokenizer([image_tokens], add_special_tokens=False, return_tensors=None)
+                for k in _text_encoding:
+                    text_encoding[k] = list(map(concat, image_token_encoding[k], _text_encoding[k]))
```
Yes, same thing, will modify to a readable version

zucchini-nlp commented on src/transformers/models/llava_next_video/diff_llava_next_video.py:
```diff
@@ -375,90 +383,119 @@ def forward(
+                    "with `processor.patch_size = {{patch_size}}` and processor.vision_feature_select_strategy = {{vision_feature_select_strategy}}`. "
+                    "Using processors without these attributes in the config is deprecated and will throw an error in v4.44."
+                )
+                if input_ids.shape[1] != 1:
+                    if image_features is not None:
+                        inputs_embeds = inputs_embeds.to(image_features.dtype)
+                        (
+                            inputs_embeds,
+                            attention_mask,
+                            position_ids,
+                            labels,
+                            input_ids,
+                        ) = self._merge_input_ids_with_image_features(
+                            image_features,
+                            feature_lens,
+                            inputs_embeds,
+                            input_ids,
+                            attention_mask,
+                            position_ids,
+                            labels=labels,
+                            image_token_index=self.config.image_token_index,
+                        )
+                    if video_features is not None:
+                        (
+                            inputs_embeds,
+                            attention_mask,
+                            position_ids,
+                            labels,
+                            input_ids,
+                        ) = self._merge_input_ids_with_image_features(
+                            video_features,
+                            video_feature_lens,
+                            inputs_embeds,
+                            input_ids,
+                            attention_mask,
+                            position_ids,
+                            labels=labels,
+                            image_token_index=self.config.video_token_index,
+                        )
```
Hmm let me see. Actually I wanted to go with one merge call for video-image models but iirc something was off with that method. I guess it was way too much logic to handle edge cases

zucchini-nlp commented on src/transformers/models/llava_next_video/processing_llava_next_video.py:
```diff
@@ -131,9 +155,62 @@ def __call__(
+
+            # videos are easier, simply get frames and multiply
+            if videos_inputs:
+                one_video = to_numpy_array(videos_inputs.get("pixel_values_videos")[0])
```
hehe, maybe it is a weird workaround. Similar to shy I used image utils to get image size, I wanted it to be agnostic for video also. But video is has an extra `time` dimension (2nd), so we can take one video and treat it like a batched image. The ` get_image_size` doesn't work with videos

I have in my TODO list to make video LLMs more standard and add some uniform helper functions, but I'll probably work on it not soon...

molbap replied:
Alright, I see! Adding what you just said in a comment would make it clearer :grin:

zucchini-nlp commented on tests/models/blip_2/test_modeling_blip_2.py:
```diff
@@ -1033,3 +1033,28 @@ def test_inference_t5_multi_accelerator(self):
+        predictions = model.generate(**inputs, do_sample=False, max_new_tokens=15)
+        generated_text = processor.batch_decode(predictions, skip_special_tokens=True)[0].strip()
+
+        # Add args to the config to trigger new logic when inputs are expanded in processing file
+        processor.num_query_tokens = model.config.num_query_tokens
+        model.resize_token_embeddings(processor.tokenizer.vocab_size, pad_to_multiple_of=64)
+        model.config.image_token_index = processor.tokenizer.vocab_size
```
yes, definitely, don't know why I ignored blip and added it for others

zucchini-nlp commented on tests/models/video_llava/test_modeling_video_llava.py:
```diff
@@ -545,3 +590,35 @@ def test_video_llava_merge_inputs_error_bug(self):
+
+    @slow
+    @require_bitsandbytes
+    def test_expansion_in_processing(self):
+        model_id = "LanguageBind/Video-LLaVA-7B-hf"
+        model = VideoLlavaForConditionalGeneration.from_pretrained(model_id, load_in_4bit=True)
+        processor = VideoLlavaProcessor.from_pretrained(model_id)
+
+        prompt = "USER: <video>Describe the video in details. ASSISTANT:"
+        video_file = hf_hub_download(
+            repo_id="raushan-testing-hf/videos-test", filename="video_demo.npy", repo_type="dataset"
+        )
+        video_file = np.load(video_file)
+
+        # check processing with expansion of inputs
+        processor.vision_feature_select_strategy = "default"
+        processor.patch_size = 14
+        inputs_expanded = processor(prompt, videos=video_file, return_tensors="pt").to(torch_device, torch.float16)
+        self.assertTrue(inputs_expanded.input_ids.shape[-1] == 2073)
+
+        # check processing without expansion of inputs (legacy behavior)
+        processor.vision_feature_select_strategy = None
+        processor.patch_size = None
+        inputs = processor(prompt, videos=video_file, return_tensors="pt").to(torch_device, torch.float16)
+        self.assertTrue(inputs.input_ids.shape[-1] == 18)
+
+        # generate exactly 20 tokens
+        output = model.generate(**inputs, min_new_tokens=20, max_new_tokens=20)
+        output_expanded = model.generate(**inputs_expanded, min_new_tokens=20, max_new_tokens=20)
+
+        # check that both inputs are handled correctly and generate the same output
+        self.assertListEqual(output_expanded[:, -20:].tolist(), output[:, -20:].tolist())
```
Yes, we could have a common testing method. Adding VLMTesterMixin which expands on GenerationTesterMixin is the second thing we need after this PR is merged

zucchini-nlp commented on tests/models/llava/test_modeling_llava.py:
```diff
@@ -186,6 +186,48 @@ def setUp(self):
+            with torch.no_grad():
+                model(**inputs)
+
+    def test_inputs_embeds_matches_input_ids(self):
+        config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
+
+        for model_class in self.all_model_classes:
+            model = model_class(config)
+            model.to(torch_device)
+            model.eval()
+
+            inputs = self._prepare_for_class(inputs_dict, model_class)
+            input_ids = inputs["input_ids"]
+            input_ids[input_ids == model.config.image_token_index] = 1  # remove image tokens
+            del inputs["input_ids"]
+            del inputs["pixel_values"]
+
+            inputs_embeds = model.get_input_embeddings()(input_ids)
+
+            with torch.no_grad():
+                out_ids = model(input_ids=input_ids, **inputs)[0]
+                out_embeds = model(inputs_embeds=inputs_embeds, **inputs)[0]
+            self.assertTrue(torch.allclose(out_embeds, out_ids))
+
```
Yes, it's about "pixel-values" or "pixel-values-videos" that have to be popped cause we can't pass them with `inputs_embeds` for VLMs. I'll check if it's generalizable enough to be added in the "common" tests, otherwise we can later override these in the VLMTesterMixin

molbap commented on src/transformers/models/llava/processing_llava.py:
```diff
@@ -107,10 +129,44 @@ def __call__(
+            and self.vision_feature_select_strategy is not None
+        ):
+            # Replace the image token with the expanded image token sequence
+            height, width = get_image_size(to_numpy_array(pixel_values[0]))
```
ah, that's true - it was to skip the `to_numpy_array` call, but got it!

zucchini-nlp committed:
remove tmp

zucchini-nlp committed:
merge `main`

zucchini-nlp committed:
codestyle

zucchini-nlp commented on src/transformers/models/paligemma/modeling_paligemma.py:
```diff
@@ -284,77 +291,43 @@ def resize_token_embeddings(self, new_num_tokens: Optional[int] = None, pad_to_m
-                (sequence_length, target_length), fill_value=min_dtype, dtype=dtype, device=device
-            )
-            if sequence_length != 1:
+        # do causal diagonal mask only if training, otherwise attend to the whole prefix
```
Just verified again, after rebasing the latest `main` that the causal masks for prev and current versions are identical. So iiuc, the mask is full in inference stage, which is how it works currently, The only thing that is masked is padding token.

In training, the image you attached is exactly how it works, with the only exception that the current token can attend to itself (as it's impl in `main` branch). The image for some reason says we shouldn't attend.

Will push soon the changes after rebasing `main`

zucchini-nlp committed:
nits

zucchini-nlp committed:
more nits

zucchini-nlp commented on tests/models/paligemma/test_modeling_paligemma.py:
```diff
@@ -87,8 +87,8 @@ def __init__(
             "use_labels": True,
-            "image_size": 30,
-            "patch_size": 2,
+            "image_size": 20,
+            "patch_size": 5,
```
Oh, just remembered why. The bigger image size/patch size we have, the longer is the sequence length of img embeddings. I thought we don't need 600+ image features for tiny models, but can bring it back for sure

zucchini-nlp committed:
remove overriding in tests

zucchini-nlp committed:
comprehension when merging video

zucchini-nlp committed:
fix-copies

zucchini-nlp committed:
revert changes for embeds test

zucchini-nlp committed:
fix tests after making comprehension

molbap commented on src/transformers/models/llava_next_video/diff_llava_next_video.py:
```diff
@@ -422,41 +422,28 @@ def forward(
-                            labels=labels,
-                            image_token_index=self.config.video_token_index,
-                        )
+                    iterator = (
+                        (image_features, feature_lens, self.config.image_token_index),
+                        (video_features, video_feature_lens, self.config.video_token_index),
+                    )
```
that's great :fire:

molbap commented:
Alright cool, taking a look soon! For the config option, a quick&dirty solution could be to do something like `_ = config.ignore_index` in the modeling?

molbap commented on src/transformers/models/paligemma/modeling_paligemma.py:
```diff
@@ -526,37 +497,24 @@ def prepare_inputs_for_generation(
+
+        # If we're in cached decoding stage, pixel values should be None because input ids do not contain special image token anymore
+        # Otherwise we need pixel values to be passed to model. NOTE: use_cache=False needs pixel_values always
+        if cache_position[0] == 0:
+            model_inputs["pixel_values"] = pixel_values
```
:+1:

molbap commented on src/transformers/models/paligemma/modeling_paligemma.py:
```diff
@@ -526,37 +497,24 @@ def prepare_inputs_for_generation(
-                "pixel_values": pixel_values,
-                "token_type_ids": token_type_ids,
-            }
+        model_inputs = self.language_model.prepare_inputs_for_generation(
+            input_ids,
+            past_key_values=past_key_values,
+            inputs_embeds=inputs_embeds,
+            attention_mask=attention_mask,
+            cache_position=cache_position,
+            **kwargs,
```
much tidier, thanks!

molbap commented on tests/models/llava/test_modeling_llava.py:
```diff
@@ -186,6 +186,49 @@ def setUp(self):
+
+    # overwrite inputs_embeds tests because we need to delete "pixel values" for LVLMs
+    # while some other models require pixel_values to be present
+    def test_inputs_embeds_matches_input_ids(self):
+        config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
+
+        for model_class in self.all_model_classes:
+            model = model_class(config)
+            model.to(torch_device)
+            model.eval()
+
+            inputs = self._prepare_for_class(inputs_dict, model_class)
+            input_ids = inputs["input_ids"]
+            del inputs["input_ids"]
+            del inputs["pixel_values"]
+
+            inputs_embeds = model.get_input_embeddings()(input_ids)
+
+            with torch.no_grad():
+                out_ids = model(input_ids=input_ids, **inputs)[0]
+                out_embeds = model(inputs_embeds=inputs_embeds, **inputs)[0]
+            self.assertTrue(torch.allclose(out_embeds, out_ids))
```
Interesting, same remark, would be worth having in common or an option that checks needed inputs for a given model to do this `del` on-demand? (nit)

zucchini-nlp replied:
yeah, I tries but seems like some models require to have both, ids and pixels, while other require only one. Will have to think about unifying Vision2Seq model tests somehow, in the scope of another PR

molbap commented on tests/models/paligemma/test_modeling_paligemma.py:
```diff
@@ -185,6 +188,49 @@ def setUp(self):
+            del inputs["input_ids"]
+            del inputs["pixel_values"]
+
+            wte = model.get_input_embeddings()
```
same remark for wte and the transformers version warning, to modify before merge!

molbap reviewed (approved):
LGTM! Some minor comments remaining but seems good

zucchini-nlp committed:
Update src/transformers/models/blip_2/processing_blip_2.py

Co-authored-by: Pablo Montalvo <39954772+molbap@users.noreply.github.com>

zucchini-nlp commented on src/transformers/models/blip_2/processing_blip_2.py:
```diff
@@ -36,20 +46,25 @@ class Blip2Processor(ProcessorMixin):
         tokenizer.return_token_type_ids = False
+        self.current_processor = image_processor
+        self.image_token = AddedToken("<image>", normalized=False, special=True)
+        tokens_to_add = {"additional_special_tokens": [self.image_token]}
+        tokenizer.add_special_tokens(tokens_to_add)
```
yeah, seems like we don't need to access token through `additional`, okey

zucchini-nlp committed:
Update src/transformers/models/blip_2/processing_blip_2.py

Co-authored-by: Pablo Montalvo <39954772+molbap@users.noreply.github.com>

zucchini-nlp commented on src/transformers/models/paligemma/modeling_paligemma.py:
```diff
@@ -285,77 +292,52 @@ def resize_token_embeddings(self, new_num_tokens: Optional[int] = None, pad_to_m
-                    padding_mask, min_dtype
-                )
-
-            final_labels = torch.full(
-                (batch_size, sequence_length), self.config.ignore_index, dtype=input_ids.dtype, device=input_ids.device
-            )
-            final_labels = torch.where(input_ids != self.pad_token_id, labels, final_labels)
```
Hmm, do we not expect users to ignore padding while preparing the labels? We can bring this back for BC but afaik the general rule is that LLMs don't mask out pad tokens in labels

zucchini-nlp replied:
returned back the masking, and added a warning that users should mask labels themselves

molbap replied:
Thanks! During training padding for uneven batches is definitely masked in labels, iiuc

zucchini-nlp committed:
more updates

zucchini-nlp committed:
fix tests

amyeroberts commented on src/transformers/models/blip_2/processing_blip_2.py:
```diff
@@ -121,9 +141,30 @@ def __call__(
             )
+
+            # if we know how many query tokens, expand text inside processor. We need this hacky manipulation
+            # because BLIP expects image tokens to be at the beginning even before BOS token
```
Nice comment :)

amyeroberts reviewed (approved):
Looks great - thanks for handling all of this!

zucchini-nlp commented on src/transformers/models/blip_2/configuration_blip_2.py:
```diff
@@ -299,7 +301,15 @@ class Blip2Config(PretrainedConfig):
+        qformer_config=None,
+        text_config=None,
+        num_query_tokens=32,
+        image_token_index=None,
```
It's token index same as in all llava models

zucchini-nlp commented:
I'll run slow tests and check everything is okey, will merge some time next week

zucchini-nlp committed:
Merge remote-tracking branch 'upstream/main' into vlm_processors

zucchini-nlp commented:
@pspdada hey! The `patch_size` and `vision_feature_select_strategy` should be taken from model config, so that `vision_feature_select_strategy = model.config.vision_feature_select_strategy` and `patch_size = model.config.vision_config.patch_size`

The message says about deprecation in v4.47 but we encountered a few things to be done so the target version will be moved a few more versions up. After the issues we encountered are fixed, I'll update official checkpoints on the hub. Until then you are welcome to use the old way (and ignore the warning) or try out the new logic by setting necessary params in the processor :)

pspdada commented:
> `vision_feature_select_strategy = model.config.vision_feature_select_strategy` and `patch_size = model.config.vision_config.patch_size`

Thank you for your response, but there is an issue. When using the following code to load the llava 1.5 model and generate with it:
```
def _create_v_1_5(self) -> tuple[LlavaForConditionalGeneration, LlavaProcessor]:
    model_name = f"llava-hf/llava-1.5-{self.model_size}-hf"

    model: LlavaForConditionalGeneration = LlavaForConditionalGeneration.from_pretrained(
        model_name,
        cache_dir=self.model_dir,
        torch_dtype=self.torch_dtype,
        device_map="auto",
        low_cpu_mem_usage=True,
        attn_implementation=attn_implementation,
    ).to(self.device).eval()

    processor: LlavaProcessor = LlavaProcessor.from_pretrained(
        model_name,
        cache_dir=self.model_dir,
        padding_side="left",
        vision_feature_select_strategy=model.config.vision_feature_select_strategy,
        patch_size=model.config.vision_config.patch_size,
    )
    print(model.config.vision_feature_select_strategy, model.config.vision_config.patch_size)

    return model, processor

def _gen(
    self,
    images: list[Image.Image],
    prompts: list[str],
    max_token: int,
    do_sample: list,
    temp: float,
    eos_token_id: list[int] | None = None,
) -> list:
    with torch.inference_mode():
        inputs = self.processor(
            images=images,
            text=prompts,
            return_tensors="pt",
            return_token_type_ids=False,
            padding=True,
        ).to(self.device, torch.float16)

        generated_ids = self.model.generate(
            **inputs,
            max_new_tokens=max_token,
            temperature=temp if do_sample else None,
            do_sample=do_sample,
            use_cache=True,
            eos_token_id=eos_token_id,
        )
        decoded_outputs: list[str] = self.processor.batch_decode(
            generated_ids,
            skip_special_tokens=False,
            clean_up_tokenization_spaces=False,
        )

    return decoded_outputs
```
The output is: `default 14`. However, when adding these two lines:
```
vision_feature_select_strategy=model.config.vision_feature_select_strategy,
patch_size=model.config.vision_config.patch_size,
```
An error occurs:
```
File "/root/llm-project/LVLM/model/llava.py", line 202, in _gen
    generated_ids = self.model.generate(
File "/root/anaconda3/envs/LVLM/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
File "/root/anaconda3/envs/LVLM/lib/python3.10/site-packages/transformers/generation/utils.py", line 2220, in generate
    result = self._sample(
File "/root/anaconda3/envs/LVLM/lib/python3.10/site-packages/transformers/generation/utils.py", line 3211, in _sample
    outputs = self(**model_inputs, return_dict=True)
File "/root/anaconda3/envs/LVLM/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
File "/root/anaconda3/envs/LVLM/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
File "/root/anaconda3/envs/LVLM/lib/python3.10/site-packages/transformers/models/llava/modeling_llava.py", line 524, in forward
    raise ValueError(
ValueError: Image features and image tokens do not match: tokens: 325440, features 576
```
If I remove those two lines, everything works fine.

pspdada commented:
An interesting phenomenon is that the original output was:
```The image features a man and a woman standing close to each other, posing for a picture.
```
However, after adding these two lines:
```
vision_feature_select_strategy=model.config.vision_feature_select_strategy,
patch_size=model.config.vision_config.patch_size,
```
the output becomes particularly strange:
```
...ge><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image> 
Please describe this image in detail and tell me what you see. ASSISTANT: The image features a man and a woman standing close to each other, posing for a picture.
```

zucchini-nlp commented:
@pspdada I just opened a PR for that (https://github.com/huggingface/transformers/pull/34332). It is a totally unrelated issue and should be solved soon

pspdada commented:
> @pspdada I just opened a PR for that (#34332). It is a totally unrelated issue and should be solved soon

Thank you for your attention, and I wish for a good outcome.

pspdada commented:
> @pspdada I just opened a PR for that (#34332). It is a totally unrelated issue and should be solved soon
I've discovered a new issue after the merge of #34332. In the latest version of `transformers==4.46.2`, problems still occur when I set:
```python
vision_feature_select_strategy=model.config.vision_feature_select_strategy,
patch_size=model.config.vision_config.patch_size,
```
After setting these parameters, the result of batch inference (without truncation, directly using `batch_decode` output) is as follows:
```
['<s> USER: <image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image> \nPlease describe this image in detail and tell me what you see. ASSISTANT: The image features a man and a woman standing close to each other, posing for a picture. The man is wearing a tie, and the woman is wearing a white shirt. They are both smiling and enjoying the moment.\n\nIn the background, there is a TV mounted on the wall, and a few bottles can be seen placed around the room. There are also two other people in the scene, one on the left side and another on the right side of the image.</s>', '<pad><pad><pad><pad><pad><pad><pad><s> USER: <image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image> \nWhat is in the image? ASSISTANT: Theo, the image\n\nWhat\n\n</s><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad>']
```

However, when I remove these two lines, everything works as expected, and the result is:
```
['<s> USER: <image> \nPlease describe this image in detail and tell me what you see. ASSISTANT: The image features a man and a woman standing close to each other, posing for a picture. The man is wearing a tie, and the woman is wearing a white shirt. They are both smiling and enjoying the moment.\n\nIn the background, there is a TV mounted on the wall, and a few bottles can be seen placed around the room. There are also two other people in the scene, one on the left side and another on the right side of the image.</s>', '<pad><pad><pad><pad><pad><pad><pad><s> USER: <image> \nWhat is in the image? ASSISTANT: The image features a group of birds perched on a tree branch.</s><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad>']
```
 Shall I open a new issue about it and provide full infomation? @zucchini-nlp

zucchini-nlp commented:
@pspdada that is the expected logic for the new processing code, as we expand the text with as many "image" tokens as there will be image embeddings. When decoding the text you should pass `skip_special_tokens=False` to remove all `image` tokens

pspdada commented:
> >

But the output of the model seems to be not quite right this way. Look at my previous example; if we remove the `<Image>` and `<pad>`, what remains is:
``` python
 \nWhat is in the image? ASSISTANT: Theo, the image\n\nWhat\n\n</s>
```
If I don't set the two line, it will be:
```
\nWhat is in the image? ASSISTANT: The image features a group of birds perched on a tree branch.</s>
```

pspdada commented:
> @pspdada that is the expected logic for the new processing code, as we expand the text with as many "image" tokens as there will be image embeddings. When decoding the text you should pass `skip_special_tokens=False` to remove all `image` tokens


But the output of the model seems to be not quite right this way. Look at my previous example; if we remove the `<Image>` and `<pad>`, what remains is:
``` python
 \nWhat is in the image? ASSISTANT: Theo, the image\n\nWhat\n\n</s>
```
If I don't set the two line, it will be:
```
\nWhat is in the image? ASSISTANT: The image features a group of birds perched on a tree branch.</s>
```

zucchini-nlp commented:
@pspdada so the new logic generates gibberish when the inputs are batched? I don't see such behavior with the latest release. Can you verify if the padding side if set to left as it is recommended in [docs](https://huggingface.co/docs/transformers/en/model_doc/llava_onevision) (`processor.tokenizer.padding_side = "left"`)

pspdada commented:
> @pspdada so the new logic generates gibberish when the inputs are batched? I don't see such behavior with the latest release. Can you verify if the padding side if set to left as it is recommended in [docs](https://huggingface.co/docs/transformers/en/model_doc/llava_onevision) (`processor.tokenizer.padding_side = "left"`)

@zucchini-nlp 
Yes, I've noticed a strange phenomenon using the latest verson of vllm
The code is:
``` python
import torch
from PIL import Image
from transformers import LlavaForConditionalGeneration, LlavaProcessor

model_name = "llava-hf/llava-1.5-7b-hf"

model: LlavaForConditionalGeneration = LlavaForConditionalGeneration.from_pretrained(
    model_name,
    cache_dir="/root/llm-project/utils/models/hub",
    torch_dtype=torch.float16,
    low_cpu_mem_usage=True,
    device_map="cuda:0",
    attn_implementation="flash_attention_2",
).eval()

processor: LlavaProcessor = LlavaProcessor.from_pretrained(
    model_name,
    cache_dir="/root/llm-project/utils/models/hub",
    padding_side="left",
)
processor.tokenizer.padding_side = "left"
processor.patch_size = model.config.vision_config.patch_size
processor.vision_feature_select_strategy = model.config.vision_feature_select_strategy

image1 = Image.open("/root/llm-project/LVLM/eval/Extended_CHAIR/images/chair-500/000000006763.jpg")
image2 = Image.open("./demo/material/1.jpg")
images = [image1, image2]
conversation1: list[dict[str]] = [
    {
        "role":
            "user",
        "content": [
            {
                "type": "image"
            },
            {
                "type": "text",
                "text": "Describe this image in detail and tell me what you see."
            },
        ],
    },
]
conversation2: list[dict[str]] = [
    {
        "role": "user",
        "content": [
            {
                "type": "image"
            },
            {
                "type": "text",
                "text": "What is in the image?"
            },
        ],
    },
]
prompt1 = processor.apply_chat_template(conversation1, tokenize=False)
prompt2 = processor.apply_chat_template(conversation2, tokenize=False)
encoded_inputs: dict[str, torch.Tensor] = processor(
    images=images,
    text=[prompt1, prompt2],
    return_tensors="pt",
    return_token_type_ids=False,
    padding=True,
).to('cuda:0')

generated_ids: torch.Tensor = model.generate(
    **encoded_inputs,
    max_new_tokens=200,
    num_beams=1,
    use_cache=True,
)
decoded_outputs: list[str] = processor.batch_decode(
    generated_ids,
    skip_special_tokens=False,
    clean_up_tokenization_spaces=True,
)

print([d.rstrip('\n').strip(" ") for d in decoded_outputs])

```

The output is:
```

['<s> USER: <image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image> \nDescribe this image in detail and tell me what you see. \n\nThe image features a man and a woman standing close to each other, posing for a picture. The man is wearing a tie, and the woman is wearing a white shirt. They are both smiling and appear to be enjoying their time together.\n\nIn the background, there is a TV mounted on the wall, and a few bottles can be seen placed around the room. There are also two other people in the scene, one standing near the left edge of the image and the other on the right side.</s>', '<pad><pad><pad><pad><pad><pad><pad><s> USER: <image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image><image> \nWhat is in the image? \nThe\n\n\nThe\n\n</s><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad>']
```
